design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+
+While range queries may target multiple shards, expanding to generic multi-shard queries will add significant flexibility to the supported interactive query forms. Multi-shard queries should be able to target any selection of RocksDB rows via comparisons performed aginst any columns:
+
+* Arbitrary WHERE clauses should be allowed.
+* Examples:
+  * `SELECT … FROM table WHERE x = ‘x’ AND y = ‘y’`
+  * `SELECT … FROM table WHERE str LIKE ‘%substr%’`
+  * `SELECT … FROM table WHERE x < y`
+
+### Tier 3 - Aggregates
+
+Aggregations are central to extracting value from datasets.
+
+While materialized views in KSQL may already be defined using aggregates, it will still be very useful to perform further aggregations over them via interactive queries. For example, users may want to maintain a materialized view summarized at the minute level but then summarize it further down to the hour level to power a frontend visualization that dynamically supplies a date range:
+
+* All aggregates supported by continuous queries should be supported by interactive queries.
+* Examples:
+  * `SELECT count(*) FROM table WHERE ROWKEY < 10`
+  * `SELECT day(hour), sum(count) FROM table GROUP BY day`
+  * `SELECT sum(count) FROM table WHERE ts >= ‘2019-01-01’ AND ts < ‘2019-01-02’`
+
+### Tier 4 - Joins
+
+Joins are also fundamental to deeply analyzing datasets and the relationships between various entities. They will also likely be the most complex to support in a clustered environment, which is why they’ve been put in a separate and final tier. As with aggregates, users may already use joins in their materialized view definitions, which may partially mitigate the cost of their absence within the context of interactive queries.
+
+For these reasons, we should probably not focus much on joins until the previous tiers have been delivered.
+
+* Without relatively robust indexing, it will probably only be initially feasible to join on row keys.
+* Examples:
+  * `SELECT … FROM t0 JOIN t1 ON t0.ROWKEY = t1.ROWKEY …`
+  * `SELECT stats_1m.count AS last_minute, stats_60m.count AS last_hour FROM stats_1m JOIN stats_60m ON stats_1m.ROWKEY = stats_1h.ROWKEY`
+
+## User experience
+
+The user experience for interactive queries should be quite simple: if a user issues a query against a KSQL table, that query should return a finite set of rows as quickly as possible.
+
+Additionally:
+
+* All interactive queries must be issued via the REST API.
+* Any interactive query can target any KSQL server.
+* Key movement should be handled transparently: if a query launches and a targeted key physically moves, KSQL should redirect the query to the appropriate KSQL server to obtain the correct result.
+* For interactive queries against windowed queries, the user may specify a time range via the `WINDOWSTART` and `WINDOWEND` psuedocolumns. For example, `SELECT * FROM windowed_aggregate WHERE ROWKEY = 'rowkey' AND WINDOWSTART >= start_ts AND WINDOWEND < end_ts`
+
+## Open questions
```

Comment:
does this proposal alter any behavior for current queries on tables? for e.g `select * from users` example here for a streaming query on a table is going to be interactive now?
https://docs.confluent.io/current/ksql/docs/developer-guide/create-a-table.html

Reply:
This question is really the crux of open question (1). If we end up adding syntax to indicate a streaming query then you'd just use that syntax to get the current behavior. If we end up identifying interactive queries as any `SELECT` on a table then that would change the existing behavior. @vinothchandar what's your preference?

Reply:
I prefer the latter i.e change current behavior (while we are young, make the current query syntax interactive and if needed introduce a special syntax for streaming query on a table later on.  IMHO this makes it much more simpler and gets us closer to making this feel very similar to any other database today.

And we just have two things, which are well-understood (out of the 4 listed in the table below) : interactive queries on tables, streaming queries on streams (which mechanisms to go from stream to table & vice-versa)

Reply:
Reducing to two things doesn't semantically work.  The semantics of a stream of data is _not_ the same as the change log of a table, so while on the surface is seems clean to just have the two, in practice it doesn't work.

The main semantic difference is the presence of tombstones in the change log of a table.  There meaning, and the way they are processed, is very different in a table vs stream.  A table treats a `null` value as a tombstone, removing the key from the table. Where as a stream would treat a `null` value as either, worse-case, something to be ignored, or best case, a row where all the value fields are `null`. 

Consider what happens if you concerted a tables changelog into a stream, filters by some predicate, and then built a new table.  Say the filter was `WHERE v0 = 10`.   When the stream encountered a `null` value it would not pass the predicate and so would be filtered out.  Hence the secondary table would not get any tombstones, and so would grow unbounded.

Hence, we need the concept of streaming queries on both tables, (or more correctly the table's changelog), and on streams, and the semantics of the two are different.